Objects/codeobject.c

```diff
@@ -151,178 +151,236 @@ validate_and_copy_tuple(PyObject *tup)
     }
-    if (intern_strings(freevars) < 0) {
+    if (intern_strings(con->varnames) < 0) {
         return NULL;
     }
-    if (intern_strings(cellvars) < 0) {
+    if (intern_strings(con->freevars) < 0) {
         return NULL;
     }
-    if (intern_string_constants(consts, NULL) < 0) {
+    if (intern_strings(con->cellvars) < 0) {
         return NULL;
     }
 
     /* Check for any inner or outer closure references */
-    assert(PyTuple_GET_SIZE(cellvars) < INT_MAX);
-    ncellvars = (int)PyTuple_GET_SIZE(cellvars);
-    assert(PyTuple_GET_SIZE(freevars) < INT_MAX);
-    nfreevars = (int)PyTuple_GET_SIZE(freevars);
-    if (!ncellvars && !nfreevars) {
-        flags |= CO_NOFREE;
+    int ncellvars = (int)PyTuple_GET_SIZE(con->cellvars);
+    if (!ncellvars && !PyTuple_GET_SIZE(con->freevars)) {
+        con->flags |= CO_NOFREE;
     } else {
-        flags &= ~CO_NOFREE;
+        con->flags &= ~CO_NOFREE;
     }
 
-    if (argcount <= nlocals && kwonlyargcount <= nlocals) {
-        /* Never overflows. */
-        total_args = (Py_ssize_t)argcount + (Py_ssize_t)kwonlyargcount +
-                      ((flags & CO_VARARGS) != 0) + ((flags & CO_VARKEYWORDS) != 0);
-    }
-    else {
-        total_args = nlocals + 1;
-    }
-    if (total_args > nlocals) {
-        PyErr_SetString(PyExc_ValueError, "code: varnames is too small");
+    PyCodeObject *co = PyObject_New(PyCodeObject, &PyCode_Type);
+    if (co == NULL) {
+        PyErr_NoMemory();
         return NULL;
     }
+    init_code(co, con);
 
     /* Create mapping between cells and arguments if needed. */
     if (ncellvars) {
-        bool used_cell2arg = false;
-        cell2arg = PyMem_NEW(Py_ssize_t, ncellvars);
-        if (cell2arg == NULL) {
-            PyErr_NoMemory();
-            return NULL;
-        }
+        int totalargs = co->co_argcount +
+                        co->co_kwonlyargcount +
+                        ((co->co_flags & CO_VARARGS) != 0) +
+                        ((co->co_flags & CO_VARKEYWORDS) != 0);
+        assert(totalargs <= co->co_nlocals);
         /* Find cells which are also arguments. */
-        for (i = 0; i < ncellvars; i++) {
-            Py_ssize_t j;
-            PyObject *cell = PyTuple_GET_ITEM(cellvars, i);
-            cell2arg[i] = CO_CELL_NOT_AN_ARG;
-            for (j = 0; j < total_args; j++) {
-                PyObject *arg = PyTuple_GET_ITEM(varnames, j);
-                int cmp = PyUnicode_Compare(cell, arg);
-                if (cmp == -1 && PyErr_Occurred()) {
-                    PyMem_Free(cell2arg);
-                    return NULL;
-                }
+        for (int i = 0; i < ncellvars; i++) {
+            PyObject *cellname = PyTuple_GET_ITEM(co->co_cellvars, i);
+            for (int j = 0; j < totalargs; j++) {
+                PyObject *argname = PyTuple_GET_ITEM(co->co_varnames, j);
+                int cmp = PyUnicode_Compare(cellname, argname);
+                assert(cmp != -1 || !PyErr_Occurred());
```

Comment:
I'm not sure what you are doing here.  
Are you asserting that an exception cannot occur? In which case `assert(!PyErr_Occurred())` would suffice.
If an exception can occur, you need to exit. This code seems to ignore the exception.


Can you add a comment explaining why the above assertion must hold.

Reply:
We don't expect an exception.  The assert is more of a just-in-case.

Reply:
Could you add a comment explaining why it can't.
I think an exception is guaranteed not to occur if both `cellname` and `argname` are readied unicode strings.
Is it guaranteed that both `cellname` and `argname` are readied unicode strings?